README.md

```diff
@@ -1,22 +1,113 @@
+
+This codebase uses Python 3.6. The main binary dependencies are MuJoCo (version 1.3.1 for
+`modelfree` and 1.5.0 for `aprl`). You may also need to install some other libraries,
+such as OpenMPI.
+
+There are two different parts of the codebase, `src/aprl` and `src/modelfree` with mutually
+exclusive requirements (due to different Gym and MuJoCo versions). You can create a
+virtual environment for each by running `ci/build_venv.sh` with either `aprl` or `modelfree`.
+Finally, inside the relevant virtual environment run `pip install -e .` to install an editable
+version of the package.
+
+# Reproducing Results
+
+For all of these experiments, you should be in the `modelfree` virtual environment:
+
+  ```bash
+  source ./modelfreevenv/bin/activate
+  ```
+
+We use [Sacred](https://github.com/IDSIA/sacred) for experiment configuration.
+
+## Training adversarial policies
+
+`modelfree.train` trains a single adversarial policy. `modelfree.multi.train` trains multiple
+adversarial policies, using Ray (see below) for parallelization.
+
+To replicate the results in the paper (there may be slight differences due to randomness not
+captured in the seeding), run `python -m modelfree.multi.train with paper`. To run the
+hyperparameter sweep, run `python -m modelfree.multi.train with hyper`.
+
+You can find results from our training run on s3://adversarial-policies-public/multi_train/paper.
+This includes TensorBoard logs, final model weights, checkpoints, and individual policy configs.
+Run `experiments/pull_public_s3.sh` to sync this and other data to `data/aws-public/`.
+
+## Evaluating adversarial policies
+
+`modelfree.score_agent` evaluates a pair of policies, for example an adversary and a victim.
+It outputs the win rate for each agent and the number of ties. It can also render to the screen
+or produce videos.
+
+We similarly use `modelfree.multi.score` to evaluate multiple pairs of policies in parallel.
+To reproduce all the evaluations used in the paper, run the following bash scripts, which call
+`modelfree.multi.score` internally:
+  - `experiments/modelfree/baselines.sh`: fixed baselines (no adversarial policies).
+  - `experiments/modelfree/attack_transfers.sh <path-to-trained-adversaries>`. To use our
+     pre-trained policies, use the path `data/aws-public/multi_train/paper/20190429_011349`
+     after syncing against S3.
+
+## Visualizing Results
+
+Most of the visualization code lives in the `modelfree.visualize` package. To reproduce the figures
+in the paper, use `paper_config`; for those in the appendix, use `supplementary_config`. So:
+
+```bash
+  python -m modelfree.visualize.scores with paper_config  # heatmaps in the paper
+  python -m modelfree.visualize.training with supplementary_config  # training curves in appendix
+```
+
+To re-generate all the videos, use `modelfree.visualize.make_videos`. We would recommend running
+in Docker using `Xdummy` for this: there are rendering issues with many graphics drivers.
```

Comment:
If it's just a one or two line example, let's add here how one would cause make_videos to run with Xdummy rather than the default rendering system; I don't actually have cached knowledge of how to do this, and I suspect it won't be clear to some others how to make this happen .

Reply:
Docker should do this automatically (although Xdummy is unreliable so sometimes I have to restart it). Clarified this.